### PR TITLE
Bootstrap command refactoring: port remaining commands with access to `Build` (step 6)

### PR DESCRIPTION
**Pull request by Kobzol**
Continuation of https://github.com/rust-lang/rust/pull/127450.

This PR ports commands in bootstrap that can easily get access to `Build(er)` to `BootstrapCommand`. After this PR, everything that can access `Build(er)` should be using the new API.

Statistics of `bootstrap` code (ignoring `src/bin/<shims>`) after this PR:
```
7 usages of `Command::new`
69 usages of `command()` (new API)
 - out of that: 16 usages of `as_command_mut()` (new API, but accesses the inner command) 
```

Tracking issue: https://github.com/rust-lang/rust/issues/126819

r? @onur-ozkan

### CODE REVIEW

**Comment by Kobzol** [comment]
Rebased to fix conflict.

File: src/bootstrap/src/lib.rs [hunk 1]
```diff
@@ -490,18 +490,30 @@ impl Build {
...
+        let submodule_git = || {
+            let mut cmd = helpers::git(Some(&absolute_path)).capture_stdout();
+            cmd.run_always();
+            cmd
+        };
 
         // Determine commit checked out in submodule.
-        let checked_out_hash = output(submodule_git().args(["rev-parse", "HEAD"]).as_command_mut());
+        let checked_out_hash =
+            submodule_git().run_always().args(["rev-parse", "HEAD"]).run(self).stdout();
```
Comment (fbstj) [inline src/bootstrap/src/lib.rs:507]:
```suggestion
            submodule_git().args(["rev-parse", "HEAD"]).run(self).stdout();
```
the `run_always()` calls on `submodule_git()` are redundant?

Comment (Kobzol) [reply src/bootstrap/src/lib.rs]: Yes, this was a leftover. Good catch, thanks! Fixed.

File: src/bootstrap/src/utils/exec.rs [hunk 2]
```diff
@@ -286,3 +287,15 @@ impl From<Output> for CommandOutput {
...
+pub fn force_coloring_in_ci(ci_env: CiEnv, cmd: &mut BootstrapCommand) {
+    if ci_env != CiEnv::None {
+        // Due to use of stamp/docker, the output stream of bootstrap is not
+        // a TTY in CI, so coloring is by-default turned off.
+        // The explicit `TERM=xterm` environment is needed for
+        // `--color always` to actually work. This env var was lost when
+        // compiling through the Makefile. Very strange.
+        cmd.env("TERM", "xterm").args(["--color", "always"]);
+    }
+}
```
Comment (onur-ozkan) [inline src/bootstrap/src/utils/exec.rs:301]: What about moving this under `impl BootstrapCommand` ?

**Review by onur-ozkan — COMMENTED** [review]
just one nit suggestion; lgtm otherwise

File: src/bootstrap/src/utils/exec.rs (hunk 2 above)
Comment (Kobzol) [reply src/bootstrap/src/utils/exec.rs]: Good idea, done that.

File: src/bootstrap/src/lib.rs [hunk 3]
```diff
@@ -522,21 +534,16 @@ impl Build {
...
-                    None
-                }
-            };
+            let current_branch = helpers::git(Some(&self.src))
+                .capture_stdout()
+                .run_always()
+                .args(["symbolic-ref", "--short", "HEAD"])
+                .run(self)
+                .stdout_if_ok()
+                .map(|s| s.trim().to_owned());
```
Comment (lukas-code) [inline src/bootstrap/src/lib.rs:543]: This broke working on a detached HEAD:

<details>
<summary>error message</summary>

```
Updating submodule src/doc/book
fatal: ref HEAD is not a symbolic ref

Command cd "/data/code/rust" && env -u GIT_ALTERNATE_OBJECT_DIRECTORIES -u GIT_DIR -u GIT_INDEX_FILE -u GIT_OBJECT_DIRECTORY -u GIT_WORK_TREE "git" "symbolic-ref" "--short" "HEAD" (failure_mode=Exit, stdout_mode=Capture, stderr_mode=Print) did not execute successfully.
Expected success, got exit status: 128
Created at: src/lib.rs:536:34
Executed at: src/lib.rs:540:18

STDOUT ----


Build completed unsuccessfully in 0:00:00
```

</details>

I think we need `.allow_failure()` here.

Comment (Kobzol) [reply src/bootstrap/src/lib.rs:543]: You're right, should be fixed by https://github.com/rust-lang/rust/pull/127919.

**Pull request merged**